README.md
# Laravel Package to Enable Eloquent "Strict Mode" and Further Safety Methods

![Laravel Strict Mode](art/laravel-strict-mode.png)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/kainiklas/laravel-strict-mode.svg?style=flat-square)](https://packagist.org/packages/kainiklas/laravel-strict-mode)
[![GitHub Tests Action Status](https://img.shields.io/github/actions/workflow/status/kainiklas/laravel-strict-mode/run-tests.yml?branch=main)](https://github.com/kainiklas/laravel-strict-mode/actions?query=workflow%3Arun-tests+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/kainiklas/laravel-strict-mode.svg?style=flat-square)](https://packagist.org/packages/kainiklas/laravel-strict-mode)

Enables the following configurable eloquent "strict mode" features:

- Prevent Lazy Loading (N+1)
    - Non-production: Throws an `Illuminate\Database\LazyLoadingViolationException` exception
    - Production: Writes a warning into the logs
- Prevent defaulting to NULL when using a model's attribute that hasn't been fetched from the DB or doesn't exist on the model
    - Throws an `Illuminate\Database\Eloquent\MissingAttributeException` exception (all environments)
    - Alternatively: You can set the config to write a warning to the log
- Prevent loosing attributes when creating or updating models because of missing attributes in the $fillable array
    - Throws an `Illuminate\Database\Eloquent\MassAssignmentException` exception (all environments)
    - Alternatively: You can set the config to write a warning to the log

Enables the following configurable safety methods:

- Long-running command monitoring
    - Writes a warning into the logs including the command name, user and duration which took longer than the specified threshold (default: 5000ms)
- Long-running requests monitoring
    - Writes a warning into the logs including the request url, user and duration which took longer than the specified threshold (default: 5000ms)
- Long-running single DB query monitoring
    - Writes an info into the logs including the SQL query and duration which took longer than the specified threshold (default: 1000ms)
- Long-running DB connection monitoring
    - Writes an info into the logs including the DB connection name which took longer than the specified threshold (default: 2000ms)
- Memory Heap Size monitoring
    - Writes a warning into the logs including when the memory heap size exceeds the specified threshold (default: 50MB)

## Installation

**Laravel Version ^10.0 required.**

You can install the package via composer:

```bash
composer require kainiklas/laravel-strict-mode
```

Optionally, you can publish the config file. 

```bash
php artisan vendor:publish --tag="laravel-strict-mode-config"
```

You can influence the behaviour with environment variables.

This is the contents of the published config file.

```php
return [

    /**
     * Throws Illuminate\Database\LazyLoadingViolationException if model is lazy loaded.
     * Exception is only thrown if log_lazy_loading is set to false.
     */
    'prevent_lazy_loading' => env(
        'PREVENT_LAZY_LOADING',
        true,
    ),

    /**
     * Lazy Loading violation is logged. No exception is thrown.
     * Only works, if prevent_lazy_loading is true.
     */
    'log_lazy_loading' => env(
        'LOG_LAZY_LOADING',
        env('APP_ENV') == 'production'
    ),

    /**
     * Prevent non-fillable attributes from being silently discarded.
     * Instead, throws Illuminate\Database\Eloquent\MassAssignmentException.
     * Exception is only thrown if log_prevent_silently_discarding_attributes is false.
     */
    'prevent_silently_discarding_attributes' => env(
        'PREVENT_SILENTLY_DISCARDING_ATTRIBUTES',
        true
    ),

    /**
     * Log warning Illuminate\Database\Eloquent\MassAssignmentException
     * instead of throwing the exception.
     * Only works if prevent_silently_discarding_attributes is true.
     */
    'log_prevent_silently_discarding_attributes' => env(
        'LOG_PREVENT_SILENTLY_DISCARDING_ATTRIBUTES',
        false
    ),

    /**
     * If activated an Illuminate\Database\Eloquent\MissingAttributeException
     * is thrown whenever an attribute is accessed which is not present in the model,
     * instead of falling back to NULL.
     *
     * Exception is only thrown if log_prevent_accessing_missing_attributes is false.
     */
    'prevent_accessing_missing_attributes' => env(
        'PREVENT_ACCESSING_MISSING_ATTRIBUTES',
        true
    ),

    /**
     * Log warning Illuminate\Database\Eloquent\MissingAttributeException
     * instead of throwing the exception.
     * Only works if prevent_accessing_missing_attributes is true.
     */
    'log_prevent_accessing_missing_attributes' => env(
        'LOG_PREVENT_ACCESSING_MISSING_ATTRIBUTES',
        false
    ),

    /**
     * Logs a warning if a command runs longer than the specified threshold.
     */
    'log_long_running_command' => env(
        'LOG_LONG_RUNNING_COMMAND',
        true
    ),

    /**
     * Threshold for long running command in milliseconds [ms].
     */
    'log_long_running_command_threshold' => env(
        'LOG_LONG_RUNNING_COMMAND_THRESHOLD',
        5000 // [ms]
    ),

    /**
     * Logs a warning if a HTTP request runs longer than the specified threshold.
     */
    'log_long_running_request' => env(
        'LOG_LONG_RUNNING_REQUEST',
        true
    ),

    /**
     * Threshold for long running HTTP request in milliseconds [ms].
     */
    'log_long_running_request_threshold' => env(
        'LOG_LONG_RUNNING_REQUEST_THRESHOLD',
        5000 // [ms]
    ),

    /**
     * Logs a warning if a DB connection runs longer than the specified threshold.
     */
    'log_long_running_total_db_query' => env(
        'LOG_LONG_RUNNING_TOTAL_DB_QUERY',
        true
    ),

    /**
     * Threshold for long running db connection in milliseconds [ms].
     */
    'log_long_running_total_db_query_threshold' => env(
        'LOG_LONG_RUNNING_TOTAL_DB_QUERY_THRESHOLD',
        2000 // [ms]
    ),

    /**
     * Logs a warning if a single DB Query runs longer than the specified threshold.
     */
    'log_long_running_single_db_query' => env(
        'LOG_LONG_RUNNING_SINGLE_DB_QUERY',
        true
    ),

    /**
     * Threshold for long running single DB Query in milliseconds [ms].
     */
    'log_long_running_single_db_query_threshold' => env(
        'LOG_LONG_RUNNING_SINGLE_DB_QUERY_THRESHOLD',
        1000 // [ms]
    ),

    /**
     * Logs a warning if a request cycle consumed more memory than the specified threshold.
     */
    'log_memory_heap_size' => env(
        'LOG_MEMORY_HEAP_SIZE',
        true
    ),

    /**
     * Threshold for memory heap size in Megabytes [MB].
     */
    'log_memory_heap_size_threshold' => env(
        'LOG_MEMORY_HEAP_SIZE_THRESHOLD',
        50 // [MB]
    ),

];
```


## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Acknowledgements

This package is based on the following article:
- [Laravel's safety mechanisms](https://planetscale.com/blog/laravels-safety-mechanisms#model-strictness) by [Aaron Francis](https://github.com/aarondfrancis)

## Credits

- [Kai Niklas](https://github.com/kainiklas)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-strict-mode` will be documented in this file.

## v2.0.0 - 2023-04-06

### Breaking Change:

- Minimum supported Version is Laravel ^10.0

### Added

- Monitor Memory Heap Size

## v1.0.2 - 2022-10-26

### Fixes

- Set correct minimum Laravel version

## v1.0.1 - 2022-10-26

- Update dependency versions

## v1.0.0 - 2022-10-26

### Added

- Ability to enable and disable all features
- Option to Log MissingAttributeException and MassAssignmentException instead of throwing an Exception
- Test cases for all features

### Breaking Changes

- Naming of config and environment variables: Please check your config and environment variables

## 0.1.1 - 2022-10-20

- Update documentation
- Defined minimum required Laravel Version ^9.35

## 0.1.0 - 2022-10-20

Initial Version


tests/Pest.php
<?php

use Kainiklas\LaravelStrictMode\Tests\TestCaseException;
use Kainiklas\LaravelStrictMode\Tests\TestCaseLogging;
use Kainiklas\LaravelStrictMode\Tests\TestCaseNotLogging;

uses(TestCaseException::class)
  // ->beforeEach(fn() => dump('beforeException'))
->in('Exception');

uses(TestCaseLogging::class)
// ->beforeEach(fn() => dump('beforeLogging'))
->in('Logging');

uses(TestCaseNotLogging::class)
// ->beforeEach(fn() => dump('beforeNotLogging'))
->in('NotLogging');

uses(TestCaseLogging::class)
// ->beforeEach(fn() => dump('beforeConsole'))
->in('Console');


tests/TestCase.php
<?php

namespace Kainiklas\LaravelStrictMode\Tests;

use Kainiklas\LaravelStrictMode\LaravelStrictModeServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelStrictModeServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
    }
}


tests/BaseTestCase.php
<?php

namespace Kainiklas\LaravelStrictMode\Tests;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;
use Kainiklas\LaravelStrictMode\LaravelStrictModeServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class BaseTestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        Schema::create('test_model1', function (Blueprint $table) {
            $table->increments('id');
        });

        Schema::create('test_model2', function (Blueprint $table) {
            $table->increments('id');
            $table->foreignId('model_1_id');
        });
    }

    protected function tearDown(): void
    {
        Model::handleLazyLoadingViolationUsing(null);
        Model::handleDiscardedAttributeViolationUsing(null);
        Model::handleMissingAttributeViolationUsing(null);
        parent::tearDown();
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelStrictModeServiceProvider::class,
        ];
    }

    protected function getEnvironmentSetUp($app)
    {
        config()->set('app.debug', 'true');
        config()->set('database.default', 'testbench');
        config()->set('database.connections.testbench', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }
}


src/Http/Middleware/MemoryHeap.php
<?php

namespace Kainiklas\LaravelStrictMode\Http\Middleware;

use Closure;
use Illuminate\Support\Facades\Log;

class MemoryHeap
{
    /**
     * Nothing to do here. Just pass-through.
     */
    public function handle($request, Closure $next)
    {
        return $next($request);
    }

    /**
     * This method is called after the response has been sent to the browser.
     * Log memory heap size in MB.
     */
    public function terminate($request, $response)
    {
        $heapsize = memory_get_peak_usage(true) / 1024 / 1024;

        if (
            config('strict-mode.log_memory_heap_size') &&
            $heapsize > config('strict-mode.log_memory_heap_size_threshold')
        ) {
            Log::warning('Memory heap size exceeded threshold.', [
                'threshold' => config('strict-mode.log_memory_heap_size_threshold'),
                'heapsize' => $heapsize.'[MB]',
            ]);
        }
    }
}


src/LaravelStrictModeServiceProvider.php
<?php

namespace Kainiklas\LaravelStrictMode;

use Illuminate\Contracts\Console\Kernel as ConsoleKernel;
use Illuminate\Contracts\Http\Kernel as HttpKernel;
use Illuminate\Database\Connection;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Kainiklas\LaravelStrictMode\Http\Middleware\MemoryHeap;
use Spatie\LaravelPackageTools\Package;
use Spatie\LaravelPackageTools\PackageServiceProvider;
use Symfony\Component\Console\Input\InputInterface;

class LaravelStrictModeServiceProvider extends PackageServiceProvider
{
    public function configurePackage(Package $package): void
    {
        /*
         * This class is a Package Service Provider
         *
         * More info: https://github.com/spatie/laravel-package-tools
         */
        $package
            ->name('laravel-strict-mode')
            ->hasConfigFile();
    }

    public function boot()
    {
        // strict mode features
        $this->handleLazyLoading();
        $this->handlePreventSilentlyDiscardingAttributes();
        $this->handlePreventAccessingMissingAttributes();

        $this->handleLongRunningDBQueries();
        $this->handleSlowCommands();
        $this->handleSlowRequests();

        $kernel = app(HttpKernel::class);
        $kernel->pushMiddleware(MemoryHeap::class);
    }

    /**
     * Reference: https://github.com/laravel/framework/pull/37363
     */
    private function handleLazyLoading()
    {
        // lazy loading => avoid N+1 queries and throw exception
        Model::preventLazyLoading(config('strict-mode.prevent_lazy_loading'));

        // lazy loading => log
        if (config('strict-mode.log_lazy_loading')) {
            Model::handleLazyLoadingViolationUsing(function ($model, $relation) {
                $class = get_class($model);

                Log::warning("Attempted to lazy load [{$relation}] on model [{$class}].");
            });
        }
    }

    /**
     * References:
     *   https://github.com/laravel/framework/pull/44664
     *   https://github.com/laravel/framework/pull/43893
     */
    private function handlePreventSilentlyDiscardingAttributes()
    {
        Model::preventSilentlyDiscardingAttributes(config('strict-mode.prevent_silently_discarding_attributes'));

        if (config('strict-mode.log_prevent_silently_discarding_attributes')) {
            Model::handleDiscardedAttributeViolationUsing(function (Model $model, $keys) {
                $class = get_class($model);
                $attributes = implode(', ', $keys);

                Log::warning("Attempted to mass assign [{$attributes}] on [{$class}].");
            });
        }
    }

    /**
     * References:
     *   https://github.com/laravel/framework/pull/44664
     *   https://github.com/laravel/framework/pull/44283
     */
    private function handlePreventAccessingMissingAttributes()
    {
        Model::preventAccessingMissingAttributes(config('strict-mode.prevent_accessing_missing_attributes'));

        if (config('strict-mode.log_prevent_accessing_missing_attributes')) {
            Model::handleMissingAttributeViolationUsing(function (Model $model, $attribute) {
                $class = get_class($model);

                Log::warning("Attempted to access [{$attribute}] on model [{$class}].");
            });
        }
    }

    /**
     * Reference: https://github.com/laravel/framework/pull/42744
     */
    private function handleLongRunningDBQueries()
    {
        // Long Running DB Query (Total)
        if (config('strict-mode.log_long_running_total_db_query')) {
            DB::whenQueryingForLongerThan(
                config('strict-mode.log_long_running_total_db_query_threshold'),
                function (Connection $connection) {
                    Log::info('Database queries exceeded the defined threshold.', [
                        'connection' => $connection->getName(),
                        'duration' => $connection->totalQueryDuration().'[ms]',
                        'threshold' => config('strict-mode.log_long_running_total_db_query_threshold').'[ms]',
                    ]);
                }
            );
        }

        // Long Running DB Query (Single)
        if (config('strict-mode.log_long_running_single_db_query')) {
            DB::listen(function ($query) {
                if ($query->time > config('strict-mode.log_long_running_single_db_query_threshold')) {
                    Log::info('An individual database query exceeded the defined threshold.', [
                        'sql' => $query->sql,
                        'duration' => $query->time.'[ms]',
                        'threshold' => config('strict-mode.log_long_running_single_db_query_threshold').'[ms]',
                    ]);
                }
            });
        }
    }

    /**
     * Reference: https://github.com/laravel/framework/pull/44125
     */
    private function handleSlowCommands()
    {
        if (! config('strict-mode.log_long_running_command')) {
            return;
        }

        if ($this->app->runningInConsole()) {
            $this->app[ConsoleKernel::class]->whenCommandLifecycleIsLongerThan(
                config('strict-mode.log_long_running_command_threshold'),
                function (Carbon $startedAt, InputInterface $input, $status) {
                    Log::warning('A command took longer than the defined threshold.', [
                        'command' => $input->getArgument('command'),
                        'duration' => $startedAt->floatDiffInSeconds() * 1000 .'[ms]',
                        'threshold' => config('strict-mode.log_long_running_command_threshold').'[ms]',
                    ]);
                }
            );
        }
    }

    /**
     * Reference: https://github.com/laravel/framework/pull/44122
     */
    private function handleSlowRequests()
    {
        if (! config('strict-mode.log_long_running_request')) {
            return;
        }

        if (! $this->app->runningInConsole() || $this->app->runningUnitTests()) {
            $this->app[HttpKernel::class]->whenRequestLifecycleIsLongerThan(
                config('strict-mode.log_long_running_request_threshold'),
                function (Carbon $startedAt, Request $request, Response $response) {
                    Log::warning('A request took longer than the defined threshold.', [
                        'user' => $request->user()?->id,
                        'url' => $request->fullUrl(),
                        'duration' => $startedAt->floatDiffInSeconds().'[s]',
                        'threshold' => config('strict-mode.log_long_running_request_threshold').'[ms]',
                    ]);
                }
            );
        }
    }
}


config/strict-mode.php
<?php

// config for Kainiklas/LaravelStrictMode
return [

    /**
     * Throws Illuminate\Database\LazyLoadingViolationException if model is lazy loaded.
     * Exception is only thrown if log_lazy_loading is set to false.
     */
    'prevent_lazy_loading' => env(
        'PREVENT_LAZY_LOADING',
        true,
    ),

    /**
     * Lazy Loading violation is logged. No exception is thrown.
     * Only works, if prevent_lazy_loading is true.
     */
    'log_lazy_loading' => env(
        'LOG_LAZY_LOADING',
        env('APP_ENV') == 'production'
    ),

    /**
     * Prevent non-fillable attributes from being silently discarded.
     * Instead, throws Illuminate\Database\Eloquent\MassAssignmentException.
     * Exception is only thrown if log_prevent_silently_discarding_attributes is false.
     */
    'prevent_silently_discarding_attributes' => env(
        'PREVENT_SILENTLY_DISCARDING_ATTRIBUTES',
        true
    ),

    /**
     * Log warning Illuminate\Database\Eloquent\MassAssignmentException
     * instead of throwing the exception.
     * Only works if prevent_silently_discarding_attributes is true.
     */
    'log_prevent_silently_discarding_attributes' => env(
        'LOG_PREVENT_SILENTLY_DISCARDING_ATTRIBUTES',
        false
    ),

    /**
     * If activated an Illuminate\Database\Eloquent\MissingAttributeException
     * is thrown whenever an attribute is accessed which is not present in the model,
     * instead of falling back to NULL.
     *
     * Exception is only thrown if log_prevent_accessing_missing_attributes is false.
     */
    'prevent_accessing_missing_attributes' => env(
        'PREVENT_ACCESSING_MISSING_ATTRIBUTES',
        true
    ),

    /**
     * Log warning Illuminate\Database\Eloquent\MissingAttributeException
     * instead of throwing the exception.
     * Only works if prevent_accessing_missing_attributes is true.
     */
    'log_prevent_accessing_missing_attributes' => env(
        'LOG_PREVENT_ACCESSING_MISSING_ATTRIBUTES',
        false
    ),

    /**
     * Logs a warning if a command runs longer than the specified threshold.
     */
    'log_long_running_command' => env(
        'LOG_LONG_RUNNING_COMMAND',
        true
    ),

    /**
     * Threshold for long running command in milliseconds [ms].
     */
    'log_long_running_command_threshold' => env(
        'LOG_LONG_RUNNING_COMMAND_THRESHOLD',
        5000 // [ms]
    ),

    /**
     * Logs a warning if a HTTP request runs longer than the specified threshold.
     */
    'log_long_running_request' => env(
        'LOG_LONG_RUNNING_REQUEST',
        true
    ),

    /**
     * Threshold for long running HTTP request in milliseconds [ms].
     */
    'log_long_running_request_threshold' => env(
        'LOG_LONG_RUNNING_REQUEST_THRESHOLD',
        5000 // [ms]
    ),

    /**
     * Logs a warning if a DB connection runs longer than the specified threshold.
     */
    'log_long_running_total_db_query' => env(
        'LOG_LONG_RUNNING_TOTAL_DB_QUERY',
        true
    ),

    /**
     * Threshold for long running db connection in milliseconds [ms].
     */
    'log_long_running_total_db_query_threshold' => env(
        'LOG_LONG_RUNNING_TOTAL_DB_QUERY_THRESHOLD',
        2000 // [ms]
    ),

    /**
     * Logs a warning if a single DB Query runs longer than the specified threshold.
     */
    'log_long_running_single_db_query' => env(
        'LOG_LONG_RUNNING_SINGLE_DB_QUERY',
        true
    ),

    /**
     * Threshold for long running single DB Query in milliseconds [ms].
     */
    'log_long_running_single_db_query_threshold' => env(
        'LOG_LONG_RUNNING_SINGLE_DB_QUERY_THRESHOLD',
        1000 // [ms]
    ),

    /**
     * Logs a warning if a request cycle consumed more memory than the specified threshold.
     */
    'log_memory_heap_size' => env(
        'LOG_MEMORY_HEAP_SIZE',
        true
    ),

    /**
     * Threshold for memory heap size in Megabytes [MB].
     */
    'log_memory_heap_size_threshold' => env(
        'LOG_MEMORY_HEAP_SIZE_THRESHOLD',
        50 // [MB]
    ),

];
